yb-voyager/src/utils/sqlname/sqlname.go

```diff
@@ -0,0 +1,230 @@
+			Unquoted:  unquote(objectName, YUGABYTE),
+			MinQuoted: minQuote(objectName, YUGABYTE),
+		},
+		SchemaName: Identifier{
+			Quoted:    `"` + schemaName + `"`,
+			Unquoted:  schemaName,
+			MinQuoted: schemaName,
+		},
+		Qualified: Identifier{
+			Quoted:    schemaName + "." + quote(objectName, YUGABYTE),
+			Unquoted:  schemaName + "." + unquote(objectName, YUGABYTE),
+			MinQuoted: schemaName + "." + minQuote(objectName, YUGABYTE),
+		},
+	}
+}
+
+func NewTargetNameFromQualifiedName(qualifiedName string) *TargetName {
+	parts := strings.Split(qualifiedName, ".")
+	if len(parts) != 2 {
+		panic(fmt.Sprintf("invalid qualified name: %s", qualifiedName))
+	}
+	return NewTargetName(parts[0], parts[1])
+}
+
+func NewTargetNameFromMaybeQualifiedName(qualifiedName string, defaultSchemaName string) *TargetName {
+	parts := strings.Split(qualifiedName, ".")
+	if len(parts) == 2 {
+		return NewTargetName(parts[0], parts[1])
+	} else if len(parts) == 1 {
+		return NewTargetName(defaultSchemaName, parts[0])
+	} else {
+		panic(fmt.Sprintf("invalid qualified name: %s", qualifiedName))
+	}
+}
+
+func (t *TargetName) String() string {
+	return t.Qualified.Quoted
+}
+
+func isQuoted(s string) bool {
+	// TODO: Learn the semantics of backticks in MySQL and Oracle.
+	return (s[0] == '"' && s[len(s)-1] == '"') || (s[0] == '`' && s[len(s)-1] == '`')
+}
+
+func quote(s string, dbType string) string {
+	if isQuoted(s) {
+		if s[0] == '`' && dbType == YUGABYTE {
+			return `"` + unquote(s, dbType) + `"` // `Foo` -> "Foo"
+		}
+		return s
+	}
+	switch dbType {
+	case POSTGRESQL, YUGABYTE:
+		return `"` + strings.ToLower(s) + `"`
+	case MYSQL:
+		return s // TODO - learn the semantics of quoting in MySQL.
+	case ORACLE:
+		return `"` + strings.ToUpper(s) + `"`
+	default:
+		panic("unknown source db type")
+	}
+}
+
+func unquote(s string, dbType string) string {
+	if isQuoted(s) {
+		return s[1 : len(s)-1]
+	}
+	switch dbType {
+	case POSTGRESQL, YUGABYTE:
+		return strings.ToLower(s)
+	case MYSQL:
+		return s
+	case ORACLE:
+		return strings.ToUpper(s)
+	default:
+		panic("unknown source db type")
+	}
+}
+
+func SetDifference(a, b []*SourceName) []*SourceName {
+	m := make(map[string]bool)
+	for _, x := range b {
+		m[x.String()] = true
+	}
+	var res []*SourceName
+	for _, x := range a {
+		if !m[x.String()] {
+			res = append(res, x)
+		}
+	}
+	return res
+}
+
+func minQuote(objectName, sourceDBType string) string {
+	objectName = unquote(objectName, sourceDBType)
+	switch sourceDBType {
+	case YUGABYTE, POSTGRESQL:
+		if isAllLowercase(objectName) {
+			return objectName
+		} else {
+			return `"` + objectName + `"`
+		}
+	case MYSQL:
+		return objectName
+	case ORACLE:
+		if isAllUppercase(objectName) {
+			return objectName
+		} else {
+			return `"` + objectName + `"`
+		}
+	default:
+		panic("invalid source db type")
+	}
+}
+
+func isAllUppercase(s string) bool {
+	for _, c := range s {
+		if c >= 'a' && c <= 'z' {
```

Comment:
what if there are special characters in the name, this `if` will be fail/invalid
Since we don't support special chars in objName, maybe you add a TODO comment here for future

Reply:
In fact, `isAllUppercase()` characters returns `false` only if it sees any lower-case character. Else it returns `true` (which includes names containing all capital letters or special characters).

The same applies to the `isAllLowercase()` implementation.

Reply:
yeah, misunderstood the code. Thanks